##  ACTIVIDAD

ENTRENAR EL MODELO Y CON LOS DATOS DE TESTEO VERIFICAR LOS INDICADORES DE VALIDACIÓN PARA LAS TRES CLASES  DE IRIS (INCLUYENDO LAS CURVAS)

En la presente actividad se trabajara con los datos de la base "iris", con el proposito de modelar la clasificación la especie de una flor a partir de las medidas de sus sepalos y petalos a través de un modelo de regresion logistica.

![imagen_iris.png](imagen_iris.png)

A continuación cargamos la base iris elaborada por Edgar Anderson:

In [3]:
iris = datasets.load_iris()
print(iris.keys())

dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename', 'data_module'])


Identificamos sus los nombres de las variables por medio de las cuales se caracteriza cada una de las flores.

In [4]:
iris.feature_names

['sepal length (cm)',
 'sepal width (cm)',
 'petal length (cm)',
 'petal width (cm)']

En la siguiente tabla se muestran las variables del conjunto de datos que seran útiles para modelar la clasificicación de las flores en la especie adecuada

In [5]:
import pandas as pd

# Creamos un dataframe con los datos de iris
iris_df = pd.DataFrame(data=iris.data, columns=iris.feature_names)
iris_df

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                  5.1               3.5                1.4               0.2
1                  4.9               3.0                1.4               0.2
2                  4.7               3.2                1.3               0.2
3                  4.6               3.1                1.5               0.2
4                  5.0               3.6                1.4               0.2
..                 ...               ...                ...               ...
145                6.7               3.0                5.2               2.3
146                6.3               2.5                5.0               1.9
147                6.5               3.0                5.2               2.0
148                6.2               3.4                5.4               2.3
149                5.9               3.0                5.1               1.8

[150 rows x 4 columns]

La clasificación de las anteriores 150 flores se asigna a continuación como a variable explicada o variable salida, cuyas posibles respuestas son los valores "0", "1" y "2", que respectivamente corresponden con las especies: setosa, versicolor y virginica.

In [8]:
Y = iris.target

### Serie de pandas para la variable dependiente

Y_df = pd.Series(Y, name='species')
Y_df

0      0
1      0
2      0
3      0
4      0
      ..
145    2
146    2
147    2
148    2
149    2
Name: species, Length: 150, dtype: int32

## Entrenamiento y Prueba

Para lograr el objetivo mecionado anteriormente se  tiene en cuenta una fase de entrenamiento del modelo y otra fase de prueba. 

Por ello a continuación se separan los datos en dos conjuntos: $70\%$ los datos de entrenamiento
y $30\%$ datos de prueba.

In [9]:
## Importamos train_test_split

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(iris_df, Y_df, test_size=0.3, random_state=42)

In [10]:
X_train

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
81                 5.5               2.4                3.7               1.0
133                6.3               2.8                5.1               1.5
137                6.4               3.1                5.5               1.8
75                 6.6               3.0                4.4               1.4
109                7.2               3.6                6.1               2.5
..                 ...               ...                ...               ...
71                 6.1               2.8                4.0               1.3
106                4.9               2.5                4.5               1.7
14                 5.8               4.0                1.2               0.2
92                 5.8               2.6                4.0               1.2
102                7.1               3.0                5.9               2.1

[105 rows x 4 columns]

In [11]:
y_train

81     1
133    2
137    2
75     1
109    2
      ..
71     1
106    2
14     0
92     1
102    2
Name: species, Length: 105, dtype: int32

Una vez elegidos los datos de entrenamiento, se crea el modelo que va a ser usado como medio para la adecuada clasificación de una nueva flor a partir de los datos de ancho y largo de los sepalos y los petalos de la flor.

In [12]:
modelo_log = LogisticRegression(random_state=42, max_iter=1000)
modelo_log

LogisticRegression(max_iter=1000, random_state=42)

A continuación se entrena el modelo con el $70\%$ de los datos.

In [13]:
modelo_log.fit(X_train,y_train)

LogisticRegression(max_iter=1000, random_state=42)

In [14]:
modelo_log.feature_names_in_

array(['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)',
       'petal width (cm)'], dtype=object)

Una vez se realiza el entrenamiento del modelo, esta listo para realizar clasificaciones con los datos de prueba.

In [19]:
modelo_log.predict(X_test)

array([1, 0, 2, 1, 1, 0, 1, 2, 1, 1, 2, 0, 0, 0, 0, 1, 2, 1, 1, 2, 0, 2,
       0, 2, 2, 2, 2, 2, 0, 0, 0, 0, 1, 0, 0, 2, 1, 0, 0, 0, 2, 1, 1, 0,
       0])

## Matriz de confusión

Una vez se tiene la clasificación de prueba se compara con los resultados de clasificación real de las especies, a tráves de una matriz de confusión:

In [21]:
from sklearn.metrics import confusion_matrix

confusion_matrix(y_test, modelo_log.predict(X_test))

array([[19,  0,  0],
       [ 0, 13,  0],
       [ 0,  0, 13]], dtype=int64)

In [31]:
Y_df=pd.DataFrame(y_test)
pred=pd.DataFrame(modelo_log.predict(X_test),index=X_test.index)
DF_revisión=pd.concat([Y_df,pred],axis=1)
DF_revisión


species  0
73         1  1
18         0  0
118        2  2
78         1  1
76         1  1
31         0  0
64         1  1
141        2  2
68         1  1
82         1  1
110        2  2
12         0  0
36         0  0
9          0  0
19         0  0
56         1  1
104        2  2
69         1  1
55         1  1
132        2  2
29         0  0
127        2  2
26         0  0
128        2  2
131        2  2
145        2  2
108        2  2
143        2  2
45         0  0
30         0  0
22         0  0
15         0  0
65         1  1
11         0  0
42         0  0
146        2  2
51         1  1
27         0  0
4          0  0
32         0  0
142        2  2
85         1  1
86         1  1
16         0  0
10         0  0

In [36]:
DF_revisión.columns = ['real', 'pred']
DF_revisión[DF_revisión['real'] != DF_revisión['pred']]

Empty DataFrame
Columns: [real, pred]
Index: []